### PR TITLE
[c10d] Add a soft error handling mode

### PR DESCRIPTION
**Pull request by kwen2501**
Adding new value "2" to env `NCCL_ASYNC_ERROR_HANDLING` standing for a "CleanUpOnly" error handling mode.
Comparing to `NCCL_ASYNC_ERROR_HANDLING=1`, the "CleanUpOnly" mode will just abort the collectives and NCCL communicators, and will not tear down the process.
User will have the chance to query the state of the process group (in a later PR) and abort the process group (in a later PR), and re-create the process group if needed.

### CODE REVIEW

File: torch/csrc/distributed/c10d/Utils.hpp [hunk 1]
```diff
@@ -80,16 +80,22 @@ inline bool parseEnvVarFlag(const char* envVarName) {
       TORCH_CHECK(false,
           "Invalid value for environment variable: " + std::string(envVarName));
     }
+    return val;
+  }
+  // Use -2 to represent unset state
+  return -2;
```
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/Utils.hpp:86]: nit: use a constant int?

**Review by rohan-varma — COMMENTED** [review]
LGTM - do we plan to add unittesting? Perhaps also a log indicating that communicators have been aborted and the process group is no longer in a healthy state, at least for the interim until the next PR adding the ability to query process group state is added?

**Review by rohan-varma — APPROVED** [review]

**Comment by kwen2501** [comment]
> LGTM - do we plan to add unittesting? Perhaps also a log indicating that communicators have been aborted and the process group is no longer in a healthy state, at least for the interim until the next PR adding the ability to query process group state is added?

Thanks @rohan-varma for the review!

Regarding the log you mentioned, I guess the following lines would be still be effective and log the error:

https://github.com/pytorch/pytorch/pull/84386/files#diff-c4e79e1dee3eea72b89d4471b91de8a7dae284e27bd5c32ed3e5296a561e4ee9L416-L421

File: torch/csrc/distributed/c10d/Utils.hpp (hunk 1 above)
Comment (kwen2501) [reply torch/csrc/distributed/c10d/Utils.hpp]: Fixed :)

**Comment by kwen2501** [comment]
The error seems to be caused by CI flakiness and not related to this PR.
@pytorchbot merge -g

**Pull request closed without merging**